proposals/0134-deprecating-exports-proposal.rst

```diff
@@ -6,137 +6,285 @@ Deprecating Exports
 .. ticket-url:: https://gitlab.haskell.org/ghc/ghc/-/issues/4879
 .. implemented::
 .. highlight:: haskell
-.. header:: This proposal was `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/134>`_.
+.. header:: This proposal was `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/134>`_, with `an amendment discussed here <https://github.com/ghc-proposals/ghc-proposals/pull/595>`_.
 .. contents::
 
-This is my GSoC 2018 project for Haskell.org
-
-The project should be helpful with easing the transition between different versions of code.
-This will be achieved by adding a support for a new deprecation pragma, which will allow the library writers to
-easily specify, which exports from a module are to be deprecated.
+This proposal introduces the ability to attach deprecation and warnings to
+top-level items in module export lists.
 
 Motivation
-------------
-As described in the ticket https://gitlab.haskell.org/ghc/ghc/issues/4879 there is sometimes a need to deprecate certain exports from a module.
+----------
+To help transition between different versions of libraries, it can be useful to
+deprecate re-exported identifiers, for example to signal that they will stop
+being exported in a future version, and to recommend alternatives (e.g. to import
+the identifiers from a different module, or to do something else entirely).
+
+This feature was originally requested in ticket https://gitlab.haskell.org/ghc/ghc/issues/4879.
 
 Proposed Change Specification
 -----------------------------
-The changes proposed are based on the first design option from the ticket https://gitlab.haskell.org/ghc/ghc/issues/4879
+The changes consist of the following aspects:
+
+* A new syntax for attaching warning and deprecations to exports,
 
-The syntax would be as follows:
+* The ability to decide when to emit deprecation warnings, which involves:
 
-::
+  - deciding which exported identifiers are deprecated, and
+  - deciding when to emit warnings at use-sites of imported deprecated
+    identifiers.
+
+Syntax changes
+~~~~~~~~~~~~~~
+We introduce the following syntax in module export lists: ::
 
     module Data.List
     (  ...
         {-# DEPRECATED "Exported from Data.String instead" #-}
         lines,
-        ...t
+        ...
     ) where
     ...
 
-The deprecation pragma would have an effect on an export that immediately follows it, only top level exports would be considered.
+The deprecation pragma has an effect on the export item that immediately
+follows it. Deprecation pragmas are only allowed on top-level exports items.
+This means we would reject the following: ::
 
-The semantics are also described in the aforementioned ticket as follows:
+    module M ( T ( {-# DEPRECATED "msg" #-} fld ) ) where { .. }
```

Comment:
I do wonder why this restriction is necessary. I would have thought that this would work and it does seem like something one might want to do.

Reply:
I see, line 108 explains this. It would be good to mention that this restriction does not limit expressiveness, as pointed out below.